optical/converter/tfrecord.py

```diff
@@ -0,0 +1,117 @@
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+from joblib import Parallel, delayed
+
+
+from .base import FormatSpec
+from .utils import _tf_parse_example, tf_decode_image
+
+
+class Tfrecord(FormatSpec):
+    def __init__(self, root: Union[str, os.PathLike]):
+        self.root = root
+        self._has_image_split = False
+        image_path = Path(root) / "images"
+        image_path.mkdir(parents=True, exist_ok=True)
+        self._annotation_dir = root
+        self._image_dir = image_path
+        self._splits = self._find_splits()
+        self._resolve_dataframe()
+
+    def _find_splits(self):
+        ann_splits = [x.stem for x in Path(self.root).glob("*.tfrecord")]
+
+        assert ann_splits, "directory doesn't have tfrecords"
+        if len(ann_splits) > 1:
+            self._has_image_split = True
+
+        return ann_splits
+
+    def _resolve_dataframe(self):
+        import tensorflow as tf
+
+        img_filenames = []
+        img_widths = []
+        img_heights = []
+        cls_names = []
+        x_mins = []
+        y_mins = []
+        box_widths = []
+        box_heights = []
+        splits = []
+        cls_ids = []
+        for split in self._splits:
+            tf_record = str(Path(self.root) / f"{split}.tfrecord")
+            render = tf.data.TFRecordDataset(tf_record)
+            dataset = render.map(_tf_parse_example)
+            img_dir = Path(self.root) / "images" / split
+            img_dir.mkdir(parents=True, exist_ok=True)
+            for data in dataset:
+                img_filename = data["image/filename"].numpy().decode("utf-8")
+                img_height = int(data["image/height"].numpy())
+                img_width = int(data["image/width"].numpy())
+                bbox_len = data["image/object/bbox/xmin"].shape[0]
+                for i in range(bbox_len):
+                    xmin = data["image/object/bbox/xmin"].values[i].numpy() * img_width
+                    ymin = data["image/object/bbox/ymin"].values[i].numpy() * img_height
+                    x_mins.append(xmin)
+                    y_mins.append(ymin)
+                    box_width = data["image/object/bbox/xmax"].values[i].numpy() * img_width - xmin
+                    box_height = data["image/object/bbox/ymax"].values[i].numpy() * img_height - ymin
+                    box_widths.append(box_width)
+                    box_heights.append(box_height)
+                    cls_names.append(data["image/object/class/text"].values[i].numpy().decode("utf-8"))
+                    img_filenames.append(img_filename)
+                    img_heights.append(img_height)
+                    img_widths.append(img_width)
+                    cls_ids.append(data["image/object/class/label"].values[i].numpy())
+                    splits.append(split)
+            _ = Parallel(n_jobs=-1, backend="threading")(
+                delayed(tf_decode_image)(self.root, data, split) for data in dataset
+            )
```

Comment:
please do a time benchmarking for joblib vs normal loop.

Reply:
There is a 3 sec difference between Joblib vs normal loop for whole Blood cell Dataset. Joblib being Fast